File: app/models/deal.py
import uuid
import enum
from datetime import date, datetime
from decimal import Decimal

from sqlalchemy import UUID, Date, DateTime, Enum, Index, Integer, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base, SharedMixin


class DealStage(str, enum.Enum):
    qualification = "qualification"
    discovery = "discovery"
    proposal = "proposal"
    negotiation = "negotiation"
    won = "won"
    lost = "lost"


class Deal(SharedMixin, Base):
    __tablename__ = "deal"

    name: Mapped[str] = mapped_column(String(255), nullable=False)
    entity_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    primary_contact_person_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), nullable=True
    )
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    stage: Mapped[DealStage] = mapped_column(
        Enum(DealStage, name="deal_stage_enum"),
        nullable=False,
        default=DealStage.qualification,
    )
    amount: Mapped[Decimal | None] = mapped_column(Numeric(18, 4), nullable=True)
    currency_code: Mapped[str] = mapped_column(String(3), nullable=False, default="USD")
    probability: Mapped[int | None] = mapped_column(Integer, nullable=True)
    expected_close_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    closed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    next_step: Mapped[str | None] = mapped_column(Text, nullable=True)
    lost_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Relationships
    entity: Mapped["Entity"] = relationship(  # noqa: F821
        "Entity", back_populates="deals"
    )

    __table_args__ = (
        Index("ix_deal_workspace_entity", "workspace_id", "entity_id"),
        Index("ix_deal_workspace_stage", "workspace_id", "stage"),
        Index("ix_deal_workspace_owner", "workspace_id", "owner_user_id"),
        Index("ix_deal_expected_close_date", "workspace_id", "expected_close_date"),
    )

File: src/lib/schema/deal.ts
import { DEAL_STAGES } from '@/lib/types';

import { OWNERSHIP_FIELDS, SYSTEM_SECTION, toneMap } from './helpers';
import type { FieldDef, ObjectLayout } from './types';

export const dealStageTone = toneMap({
  qualification: 'outline',
  discovery: 'outline',
  proposal: 'secondary',
  negotiation: 'secondary',
  won: 'default',
  lost: 'destructive',
});

const FIELDS = {
  name: { key: 'name', label: 'Deal name', type: 'text', required: true } as FieldDef,
  entity_id: {
    key: 'entity_id',
    label: 'Entity',
    type: 'lookup',
    required: true,
    lookup: { resource: 'entities', labelOf: (row: Record<string, unknown>) => String(row.name) },
  } as FieldDef,
  primary_contact_person_id: {
    key: 'primary_contact_person_id',
    label: 'Primary contact',
    type: 'lookup',
    lookup: {
      resource: 'persons',
      labelOf: (row: Record<string, unknown>) => [row.first_name, row.last_name].filter(Boolean).join(' '),
    },
  } as FieldDef,
  stage: { key: 'stage', label: 'Stage', type: 'select', options: DEAL_STAGES, required: true, badgeTone: dealStageTone } as FieldDef,
  amount: { key: 'amount', label: 'Amount', type: 'currency' } as FieldDef,
  probability: { key: 'probability', label: 'Probability (%)', type: 'number' } as FieldDef,
  expected_close_date: { key: 'expected_close_date', label: 'Expected close', type: 'date' } as FieldDef,
  closed_at: { key: 'closed_at', label: 'Closed at', type: 'datetime', readOnly: true } as FieldDef,
  next_step: { key: 'next_step', label: 'Next step', type: 'longtext' } as FieldDef,
  lost_reason: { key: 'lost_reason', label: 'Lost reason', type: 'longtext' } as FieldDef,
  description: { key: 'description', label: 'Description', type: 'longtext' } as FieldDef,
  notes: { key: 'notes', label: 'Internal notes', type: 'longtext' } as FieldDef,
};

export const DEAL_FIELDS = FIELDS;

export const DEAL_LAYOUT: ObjectLayout = {
  sections: [
    {
      title: 'Deal Information',
      fields: [FIELDS.name, FIELDS.entity_id, FIELDS.primary_contact_person_id, FIELDS.stage, FIELDS.amount, FIELDS.probability],
    },
    {
      title: 'Timeline',
      fields: [FIELDS.expected_close_date, FIELDS.closed_at, FIELDS.next_step, FIELDS.lost_reason],
    },
    { title: 'Ownership', fields: OWNERSHIP_FIELDS },
    { title: 'Additional Information', fields: [FIELDS.description, FIELDS.notes] },
    SYSTEM_SECTION,
  ],
};

export const DEAL_LIST_COLUMNS = [
  FIELDS.name,
  FIELDS.entity_id,
  FIELDS.stage,
  FIELDS.amount,
  FIELDS.expected_close_date,
  { ...OWNERSHIP_FIELDS[0] },
];

export const DEAL_SEARCH_FIELDS = ['name'];
